### PR TITLE
Implementing Internationalization

### LINKED ISSUE #4: Internationalization
Can this be set up for translation via POT and PO files?

### PR DESCRIPTION
**Pull request by pipix51**
This pull request fixes #4 . I have added the .pot file for translation as well as English translations (in en_GB.po and en_US.po). I have also cleaned up some of the syntax and grammatical errors in consts.py and globals.py .

### CODE REVIEW

**Comment by leonardcj** [comment]
I will pull the POT and host on our Pootle instance as soon as I can.

**Comment by quozl** [comment]
@pipix51, thanks for that, reviewed;

- [x] the file `PolariActivity.pot` says it is for Music Blocks,
- [x] the file `PolariActivity.pot` is licensed LGPLv3 but Polari Activity seems mostly to be GPLv3+; are you sure you want to license it as LGPLv3?
- [x] the file `globals.py` is not used, so you may remove it instead of changing it,
- [x] your change to `LOST_CONNECTION` in b85700d4324d6980bd5ba0d0d159932bc1145878 missed a spelling mistake, though you fixed it in the POT file, therefore translations won't be found for this message,
- [x] critical strings in the user interface, like Nick, Server, Port, Channel, Cancel, Connect!, (no topic set), Status, and Speak have not been wrapped for translation,
- [ ] regenerating the POT file using our tools fails,

```
host:~/polari-activity$ python setup.py genpot
xgettext: Non-ASCII string at twisted/conch/ssh/connection.py:399.
          Please specify the source encoding through --from-code.
ERROR - xgettext failed with return code 1.
host:~/polari-activity$ 
```

File: twisted/conch/ssh/connection.py [hunk 1]
```diff
@@ -396,7 +396,7 @@ def sendGlobalRequest(self, request, data, wantReply=0):
         """
         self.transport.sendPacket(MSG_GLOBAL_REQUEST,
                                   common.NS(request)
-                                  + (wantReply and b'\xff' or b'\x00')
+                                  + (wantReply and b'0xff' or b'0x00')
```
Comment (quozl) [inline twisted/conch/ssh/connection.py]: This change does not look right.

```
$ python
>>> b'\xff' == b'0xff'
False
>>> 
```

**Comment by pipix51** [comment]
I did some digging, the hexadecimal codes are needed by the protocol to work and cannot be changed. However, these codes are not in ASCII. Do we have to change the source encoding to another standard?

**Comment by quozl** [comment]
I don't know the internals of `twisted`, but there is a whole community for it out there somewhere, and at least we have a copy of the source here, so it should be possible to figure out.  On the face of it, I don't know why the code takes a variable, performs an `and` with a string containing 0b11111111, then performs an `or` with a string containing 0b00000000.  What is the type of `wantReply`?  Perhaps `bool`.  So the expression converts a boolean to a single character string?  What do you think?

```
>>> True and '\xff' or '\x00'
'\xff'
>>> False and '\xff' or '\x00'
'\x00'
```

**Comment by pipix51** [comment]
wantReply is indeed a boolean as indicated in the comment on line 394 in the same file. I agree with you that the expression converts a boolean to a single character string. However there is still one thing that bugs me, which is the single character strings. In ASCII, '\xff' represents "ÿ", something that does not really make sense here.

**Comment by quozl** [comment]
In [ASCII](https://en.wikipedia.org/wiki/ASCII), `ff` is usually treated as `DEL`.

I've experience in several languages, including assembler, C, and Forth.  The expression now makes a bit more sense to me.  `'\xff'` is a constant of type `str` with a single character-byte value consisting of eight bits set.  Like C's `unsigned char` or `uint_8`.

Why not use `chr()` built-in function of Python instead of string constants;

```
>>> chr(255) == '\xff'
True
>>> chr(0) == '\x00'
True
```

That may avoid the `xgettext` problem without changing the semantics of the Twisted code.

Then, with the `genpot` target working, you can add `import gettext` and `_()` calls where you need them.

**Comment by pipix51** [comment]
I added the critical strings to the .pot file and replaced the non-ASCII strings, as well as deleting the unused globals.py file.

**Comment by quozl** [comment]
Thanks, good progress, but please test your code before commit.  As at e2a5821, the activity does not start, and there is a traceback in logs.  It shows me your code was not run.

There's no need to translate for log messages; the `print` in `client.py`.

Also, translators find it slow going to handle messages with `%s` unnamed argument format strings, these are best replaced with format strings with named arguments.  `genpot` warns me about this for several lines;

* `client.py` (214, 246,  326),
* `polari_canvas.py` (270, 291).

There's also another non-ASCII string error in `userauth.py` line 220.  Please make sure you run the `genpot` target yourself to see these warnings or errors.

**Comment by quozl** [comment]
Good progress, well done!

I've checked out 087caaa, reviewed and tried to run it.  It still fails.

- [ ] please run the activity on your Linux development system, and fix the errors it reports when it fails to start,

I've checked the `genpot` target again.  It still fails.

- [ ] please install `sugar-toolkit-gtk3` on your Linux development system, and run `python setup.py genpot`, fix any reported failures, then iterate until `genpot` returns zero; your fixes may be either to source code of twisted (as you have done), or to `setup.py` to exclude twisted from the `genpot` target, your choice,

I've not been able to find your changes to the critical user interface strings.

- [x] please locate and wrap the user interface strings from the following screenshots, marked with purple dot, like Nick, Server, Port, Channel, Cancel, Connect!, (no topic set), Status, and Speak,


screenshot_ubb-from-uaa_2017-12-14_15 00 30
screenshot_ubb-from-uaa_2017-12-14_15 01 10
screenshot_ubb-from-uaa_2017-12-14_15 01 32

Thanks!

**Comment by quozl** [comment]
This is looking like the wrong way to approach the problem.  The amount of change to the twisted source makes it diverge from the upstream project.  Please research xgettext options and our `setup.py` bundlebuilder implementation in sugar-toolkit-gtk3 to see if there is a better way.

**Comment by quozl** [comment]
Take a look at https://github.com/sugarlabs/gogo/blob/master/setup.py for a method to ignore certain directories.  I'm not sure if it works with the `genpot` target, but you could read the source of bundlebuilder to check.

**Comment by pipix51** [comment]
The IGNORE_DIRS and IGNORE_FILES arguments do not seem to work with genpot as the errors still occur.

**Comment by quozl** [comment]
Thanks.  `bundlebuilder.cmd_genpot `doesn't have a feature to ignore files.  Perhaps try this;

* read [sugar3.activity.bundlebuilder documentation](https://developer.sugarlabs.org/sugar3/sugar3.activity.bundlebuilder.html),
* read [setup.py](https://github.com/sugarlabs/PolariActivity/blob/master/setup.py) and take note of how  `bundlebuilder` is run entirely from the installed sugar-toolkit-gtk3,
* read [bundlebuilder.py from sugar-toolkit-gtk3](https://github.com/sugarlabs/sugar-toolkit-gtk3/blob/master/src/sugar3/activity/bundlebuilder.py) and take note of how `bundlebuilder` works in general, and the API in particular,
* read [setup.py from tamtam](https://github.com/sugarlabs/tamtam/blob/master/setup.py) to see how another activity implements a local variant of the `bundlebuilder`,
* edit setup.py to add a local variant of `genpot`, leaving the rest of the functionality to `bundlebuilder` in the toolkit.

Unless you have any other ideas?  e.g. could the twisted source be placed in a zip bundle that Python will unpack on the fly during import?

**Comment by pipix51** [comment]
The setup.py file in tamtam also only calls bundlerbuilder.start() when genpot is called, the additional code is for other commands.

**Comment by quozl** [comment]
Yes, I know.

**Comment by quozl** [comment]
Still waiting for the activity to start without failing, how is that going?

**Comment by pipix51** [comment]
Apologies, I'm still stuck at generating the POT file for the activity...

**Comment by quozl** [comment]
That's a good start, but could you delegate most of setup.py to sugar3.activity.bundlebuilder so that when changes occur to sugar3 the changes don't need to be copied into setup.py?

**Comment by pipix51** [comment]
I have moved all of the functions except the genpot function to bundlebuilder and modified the genpot function to differentiate it from the one in the bundlebuilder. The command to get the POT file now works as I have excluded the 'twisted' directory from the list of files to be translated, however it is now `python setup.py genpot_polari` instead of `python setup.py genpot` to avoid clashing.

**Comment by quozl** [comment]
Thanks.  Very close to perfect.  But f619ee9 does fail to run the `dist_xo` target in `setup.py`, prints help instead.  Because `cmd_dist_xo` is not in `globals()`.

Just guessing, but perhaps check for `name in globals()` instead of enclosing in an exception handler?  Or locate the `bundlebuilder` function another way?  e.g. call `bundlebuilder.start` if the argument is not `genpot`?

Up to you.

**Comment by pipix51** [comment]
The command `cmd_dist_xo` can be executed now; I have added the check you mentioned for `options.command in globals()`.

**Comment by quozl** [comment]
Tested 3d2b24f.  The `dist_xo` command can be "executed" in that no error is seen, but it does not run, and there is no bundle created in the `dist` directory.

**Comment by pipix51** [comment]
I have added an if/else clause which causes the dist_xo to run, however it throws a warning which states that there is a syntax error in en_US.po . Do I delete the above .po files?

**Comment by quozl** [comment]
Thanks.  Reviewed and tested.  Much improved.

- [x] your `en_GB.po` and `en_US.po` files are not properly formatted, as if you didn't use the tools to create them; it is easy to fix the format, you might compare against `po` files in other activities; let me know if you need help with that,
- [x] we all expect to type `python setup.py genpot`, and we may automate that later, so if you can hack it to accept that instead of `genpot_polari`, all the better,
- [x] your `setup.py` now has a redundant exception handler; remember the handler was there to avoid having to check for a `globals()` member, and you've since fixed that, so at the moment any `KeyError` or `IndexError` inside `bundlebuilder` will be silently ignored,
- [ ] began detailed code review, stopped on lots of imports in your `setup.py`, so wondered if they were still needed; many flake8 errors, such as F401 imported but unused, please run `python -m flake8 setup.py` until you've resolved them.

**Comment by pipix51** [comment]
I reformatted the en_GB.po and en_US.po files and renamed genpot_polari to genpot. I also removed the try/except loop and cleaned up the code in setup.py using flake8.

**Comment by quozl** [comment]
Thanks, great work.  Rather than talk you through the other changes, I've merged your changes and pushed a72fab3a1a09e43619be472fd3b2abb38aa221b8 thus;

* `git merge pipix51/master`
* `git reset 457fbc482b422146351564fcbc1fed50bfab9cea`
* discarded the changes to `twisted`; they are no longer needed now with custom `genpot` in `setup.py`
* discarded the new file `po/PolariActivity.pot`; the file name made by `genpot` is `po/Polari.pot` which would be confusing, and there are textual changes between the two files,
* discarded the new files `po/POTFILES.in` and `po/POTFILES.skip`; as they are not needed,
* ran `genpot` to update `po/Polari.pot`,
* changed `setup.py` to simplify the call to `cmd_genpot`, avoiding the `globals()` indirection,
* changed `setup.py` to avoid calling `bundlebuilder.Config` twice for the non-`genpot` targets,
* removed wrap from `gtk3reactor.install` error, as it is not user visible,
* added `.gitignore`.

Any questions, please ask.

**Comment by quozl** [comment]
Oops, attribution error.  Did `git commit --amend --author` and `git push --force`, new commit is de88592121f7afe1361d07a42c4e570e58c8bc07

**Comment by quozl** [comment]
Testing showed missing imports, fixed in 6fc4839e9926f0593c89a8396d029802c4cf1d3d.

**Comment by pipix51** [comment]
No questions here. Thank you for your guidance!

**Pull request closed without merging**